pkg/inclusion/paths.go

```diff
@@ -0,0 +1,123 @@
+package inclusion
+
+// genSubTreeRootPath calculates the path to a given subtree root of a node, given the
+// depth and position of the node. note: the root of the tree is depth 0.
+// The following nolint can be removed after this function is used.
+//nolint:unused,deadcode
+func genSubTreeRootPath(depth int, pos uint) []bool {
+	path := make([]bool, depth)
+	for i := depth; i >= 0; i-- {
+		if (pos & (1 << i)) == 0 {
+			path = append(path, false)
+		} else {
+			path = append(path, true)
+		}
+	}
+	return path
+}
+
+// coord identifies a tree node using the depth and position
+// Depth       Position
+// 0              0
+//               / \
+//              /   \
+// 1           0     1
+//            /\     /\
+// 2         0  1   2  3
+//          /\  /\ /\  /\
+// 3       0 1 2 3 4 5 6 7
+type coord struct {
+	// depth is the typical depth of a tree, 0 being the root
+	depth uint64
+	// position is the index of a node of a given depth, 0 being the left most
+	// node
+	position uint64
+}
+
+// climb is a state transition function to simulate climbing a balanced binary
+// tree, using the current node as input and the next highest node as output.
```

Comment:
I don't think being able to climb right is a technically a precondition? or perhaps I'm just not clear on what you mean by this? the implementation that uses this code only climbs if it can, but is that different?